lib/mix/lib/mix/tasks/compile.elixir.ex

```diff
@@ -15,6 +15,35 @@ defmodule Mix.Tasks.Compile.Elixir do
   Note it is important to recompile a file's dependencies as
   there are often compile time dependencies between them.
 
+  ## `__mix_recompile__?/0`
+
+  A module may export a `__mix_recompile__?/0` function that can
+  cause the module to be recompiled using custom rules. For example,
+  `@external_resource` already adds a compile-time dependency on an
+  external file, however to depend on a _dynamic_ list of files we
+  can do:
+
+      defmodule MyModule do
+        paths = Path.wildcard("*.txt")
+        paths_hash = :erlang.md5(paths)
+
+        for path <- paths do
+          @external_resource path
+        end
+
+        def __mix_recompile__?() do
+          Path.wildcard("*.txt") |> :erlang.md5() == unquote(paths_hash)
+        end
+      end
+
+  Compiler calls `__mix_recompile__?/0` for every module being
+  compiled (or previously compiled) and thus it is very important
+  to do there as little work as possible to not slow down the
+  compilation.
+
+  If module has `@compile {:autoload, false}`, `__mix_recompile__?/0` will
```

Comment:
I don't see where the compile attribute is checked. It looks like we always call `Code.ensure_loaded?(module)`.

Reply:
I was assuming some other code was already doing that. We test this behaviour here: https://github.com/elixir-lang/elixir/pull/9919/files#diff-6e7644e6f8ea5614b076468d094aaba1R501, A and C have `def __mix_recompile__?(), do: true`, but because C is setting `@compile {:autoload, false}`, it does not get recompiled.